geomstats/sphere.py

```diff
@@ -0,0 +1,114 @@
+"""
+Computations on the n-dimensional sphere
+embedded in the (n+1)-dimensional Euclidean space.
+"""
+
+import numpy as np
+
+EPSILON = 1e-6
+
+
+def projection_to_tangent_space(ref_point, vector):
+    """
+    Project the vector v onto the tangent space:
+    T_{ref_point} S = {w | scal(w, ref_point) = 0}
+    """
+    sq_norm = np.dot(ref_point, ref_point)
+    tangent_vec = vector - np.dot(ref_point, vector) / sq_norm * ref_point
+    return tangent_vec
+
+
+def riemannian_exp(ref_point, vector, epsilon=EPSILON):
+    """
+    Compute the Riemannian exponential at point ref_point
+    of tangent vector tangent_vec wrt the metric obtained by
+    embedding of the n-dimensional sphere
+    in the (n+1)-dimensional euclidean space.
+
+    This gives a point on the n-dimensional sphere.
+
+    :param ref_point: a point on the n-dimensional sphere
+    :param vector: (n+1)-dimensional vector
+    :returns riem_exp: a point on the n-dimensional sphere
+    """
+    tangent_vec = projection_to_tangent_space(ref_point, vector)
+    norm_tangent_vec = np.linalg.norm(tangent_vec)
+
+    if norm_tangent_vec < epsilon:
+        coef_1 = (1. - norm_tangent_vec ** 2 / 2.
+                  + norm_tangent_vec ** 4 / 24.
+                  - norm_tangent_vec ** 6 / 720.
+                  + norm_tangent_vec ** 8 / 40320.)
+        coef_2 = (1. - norm_tangent_vec ** 2 / 6.
+                  + norm_tangent_vec ** 4 / 120.
+                  - norm_tangent_vec ** 6 / 5040.
+                  + norm_tangent_vec ** 8 / 362880.)
+    else:
+        coef_1 = np.cos(norm_tangent_vec)
+        coef_2 = np.sin(norm_tangent_vec) / norm_tangent_vec
+
+    riem_exp = coef_1 * ref_point + coef_2 * tangent_vec
+
+    return riem_exp
+
+
+def riemannian_log(ref_point, point, epsilon=EPSILON):
+    """
+    Compute the Riemannian logarithm at point ref_point,
+    of point wrt the metric obtained by
+    embedding of the n-dimensional sphere
+    in the (n+1)-dimensional euclidean space.
+
+    This gives a tangent vector at point ref_point.
+
+    :param ref_point: point on the n-dimensional sphere
+    :param point: point on the n-dimensional sphere
+    :returns riem_log: tangent vector at ref_point
+    """
+    norm_ref_point = np.linalg.norm(ref_point)
+    norm_point = np.linalg.norm(point)
+
+    cos_angle = np.dot(ref_point, point) / (norm_ref_point * norm_point)
+    if cos_angle > 1.0:
+        angle = 0.0
+    else:
+        angle = np.arccos(cos_angle)
+
+    if angle < epsilon:
+        coef_1 = (1. + angle ** 2 / 6.
```

Comment:
same comment as before for factorials

Reply:
These are the taylor expansion of x / sin(x) and x / tan(x), whose coefficients are computed *from* the factorials in the expansions of sin(x) and tan(x) using the computational rules on  taylor expansion's inversion (basically writing that the product of the taylor expansions will give 1 and identifying the coefficients).

Since the numerical factors are not exactly the inverse of factorials, and there will be sums of inverses of factorials, I don't think that it makes sense to make the factorials appear here.

See: https://www.wolframalpha.com/input/?i=1+%2F+(sin(x)) for the result.

Reply:
ok, can you just make it more readable - it's just pretty awful to see plain numbers like this. Maybe you could define each coefficient for each term somewhere? Maybe in an array or a dictionnary?

Something like:

TAYLOR_COEFFS = [1, 1, 7./360., 31./15120.] ...

Same for previous function actually.

This will also have the side effect of making it clear to the reader that this is a taylor expansion.

Reply:
Great idea! thanks.